### PR TITLE
Add server implementation

### PR DESCRIPTION
**Pull request by zcdirk**

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -15,8 +22,66 @@ const (
...
+		}
+		mapLock.Unlock()
+
+		go sendMessage(c, stream)
+	}
+	for !s.quit {}
+	return nil
+}
+
+func sendMessage(c chan *pb.Message, stream pb.PubSub_SubscribeServer) {
```
Comment (xinxiao) [inline main.go:72]: I assume you meant to push the message to the clients here. "listenToMessage" could be a better name.

File: main.go [hunk 2]
```diff
@@ -15,8 +22,66 @@ const (
...
+		if ok {
+			topicToChannelsMap[topic.Name] = append(channels, c)
+		} else {
+			topicToChannelsMap[topic.Name] = []chan *pb.Message{c}
+		}
+		mapLock.Unlock()
+
+		go sendMessage(c, stream)
+	}
+	for !s.quit {}
```
Comment (xinxiao) [inline main.go:68]: If we follow the idea earlier we don'e have to block the process here.

And I don't think you'll never exit this call unless the client disconnect, which could potentially be a problem.

File: main.go [hunk 3]
```diff
@@ -15,8 +22,66 @@ const (
...
+		mapLock.Lock()
+		channels, ok := topicToChannelsMap[topic.Name]
+		if ok {
+			topicToChannelsMap[topic.Name] = append(channels, c)
+		} else {
+			topicToChannelsMap[topic.Name] = []chan *pb.Message{c}
+		}
+		mapLock.Unlock()
+
+		go sendMessage(c, stream)
```
Comment (xinxiao) [inline main.go:66]: Looks like you would have to build a channel to subscriber for each topic, which means that if we a client subscribe `n` topics, you would maintain `n` channels and spawn `n` go-routines, which is quite expensive. 

Managing only one go-routine and one channel for each client could be a more elegant and approachable solution for you to consider.

File: main.go (hunk 1 above)
Comment (zcdirk) [reply main.go]: Change to listenAndRespond

File: main.go (hunk 3 above)
Comment (zcdirk) [reply main.go]: Good point

File: main.go (hunk 2 above)
Comment (zcdirk) [reply main.go]: Actually, we do need to never exit this call. I rewrite it in another way. Once we return, the client will be closed.

File: main.go [hunk 4]
```diff
@@ -15,8 +22,63 @@ const (
...
+			topicToChannelsMap[topic.Name] = []chan *pb.Message{c}
+		}
+		mapLock.Unlock()
+	}
+	return listenAndRespond(c, stream)
+}
+
+func listenAndRespond(c chan *pb.Message, stream pb.PubSub_SubscribeServer) error {
+	for {
+		select {
```
Comment (xinxiao) [inline main.go:70]: what is `select` here used for.

File: main.go [hunk 5]
```diff
@@ -15,8 +22,63 @@ const (
...
+		mapLock.Lock()
+		channels, ok := topicToChannelsMap[topic.Name]
+		if ok {
+			topicToChannelsMap[topic.Name] = append(channels, c)
+		} else {
+			topicToChannelsMap[topic.Name] = []chan *pb.Message{c}
+		}
+		mapLock.Unlock()
+	}
+	return listenAndRespond(c, stream)
```
Comment (xinxiao) [inline main.go:65]: The `error` here is not the error that might occur during the streaming process, but the error that emerged while attempting to initialize the streaming stub. The idea approach is like this:

```
go listenAndRespond(c, stream)
return nil
```

File: main.go (hunk 4 above)
Comment (zcdirk) [reply main.go]: Removed.

File: main.go (hunk 5 above)
Comment (zcdirk) [reply main.go:65]: Please note that the Subscribe service is "A server-to-client streaming RPC". Once we return, the client side will close the stream and server will no longer be able to send msg to client. 

We may later on consider changing this to a bidirectional streaming RPC.

File: main.go [hunk 6]
```diff
@@ -15,8 +22,61 @@ const (
 
 var (
 	path = flag.String("config", "", "path to pubsub server config")
+	topicToChannelsMap = make(map[string][]chan *pb.Message)
+	mapLock = new(sync.Mutex)
 )
 
+type pubsubServer struct {
+	pb.UnimplementedPubSubServer
+	quit bool
```
Comment (xinxiao) [inline main.go:31]: Do we still need quit here?

Comment (zcdirk) [reply main.go]: Good catch. Removed.

**Pull request merged**